sw/device/lib/dif/dif_pwrmgr.c

```diff
@@ -4,8 +4,563 @@
+            kDifPwrmgrDomainOptionUsbClockInActivePower |
+            kDifPwrmgrDomainOptionMainPowerInLowPower,
+    .index = PWRMGR_CONTROL_CORE_CLK_EN,
+};
+
+/**
+ * Relevant bits of the WAKEUP_EN and WAKE_INFO registers must start at `0` and
+ * be in the same order as `dif_pwrmgr_wakeup_request_source_t` constants.
+ */
+_Static_assert(kDifPwrmgrWakeupRequestSourceOne ==
+                   (1u << PWRMGR_WAKEUP_EN_EN_0),
+               "Layout of WAKEUP_EN register changed.");
+_Static_assert(kDifPwrmgrWakeupRequestSourceOne ==
+                   (1u << PWRMGR_WAKE_INFO_REASONS),
+               "Layout of WAKE_INFO register changed.");
+
+/**
+ * Relevant bits of the RESET_EN register must start at `0` and be in the same
+ * order as `dif_pwrmgr_reset_request_source_t` constants.
+ */
+_Static_assert(kDifPwrmgrResetRequestSourceOne == (1u << PWRMGR_RESET_EN_EN_0),
+               "Layout of RESET_EN register changed.");
+
+/**
+ * `dif_pwrmgr_irq_t` constants must match the corresponding generated values.
+ */
+_Static_assert(kDifPwrmgrIrqWakeup == PWRMGR_INTR_COMMON_WAKEUP,
+               "Layout of interrupt registers changed.");
+
+/**
+ * Register information for wakeup and reset requests.
+ */
+typedef struct request_reg_info {
+  ptrdiff_t write_enable_reg_offset;
+  bitfield_bit32_index_t write_enable_bit_index;
+  ptrdiff_t sources_reg_offset;
+  ptrdiff_t current_sources_reg_offset;
+  bitfield_field32_t bitfield;
+} request_reg_info_t;
+
+static const request_reg_info_t request_reg_infos[2] = {
+        [kDifPwrmgrReqTypeWakeup] =
+            {
+                .write_enable_reg_offset = PWRMGR_WAKEUP_EN_REGWEN_REG_OFFSET,
+                .write_enable_bit_index = PWRMGR_WAKEUP_EN_REGWEN_EN,
+                .sources_reg_offset = PWRMGR_WAKEUP_EN_REG_OFFSET,
+                .current_sources_reg_offset = PWRMGR_WAKE_STATUS_REG_OFFSET,
+                .bitfield =
+                    {
+                        .mask = kDifPwrmgrWakeupRequestSourceOne, .index = 0,
+                    },
+            },
+        [kDifPwrmgrReqTypeReset] =
+            {
+                .write_enable_reg_offset = PWRMGR_RESET_EN_REGWEN_REG_OFFSET,
+                .write_enable_bit_index = PWRMGR_RESET_EN_REGWEN_EN,
+                .sources_reg_offset = PWRMGR_RESET_EN_REG_OFFSET,
+                .current_sources_reg_offset = PWRMGR_RESET_STATUS_REG_OFFSET,
+                .bitfield =
+                    {
+                        .mask = kDifPwrmgrResetRequestSourceOne, .index = 0,
+                    },
+            },
+};
+
+/**
+ * Checks if a value is a valid `dif_pwrmgr_toggle_t` and converts it to `bool`.
+ */
+DIF_WARN_UNUSED_RESULT
+static bool toggle_to_bool(dif_pwrmgr_toggle_t val, bool *val_bool) {
+  switch (val) {
+    case kDifPwrmgrToggleEnabled:
+      *val_bool = true;
+      break;
+    case kDifPwrmgrToggleDisabled:
+      *val_bool = false;
+      break;
+    default:
+      return false;
+  }
+  return true;
+}
```

Comment:
Initially I had a function that only checked if the `val` was a valid toggle. Then I moved toggle -> bool conversion to that function too but wanted to keep the validity check as well. I know returning `bool` for both the return value and the out parameter can be a little bit confusing but this is a static function and consistent with the other static functions in the file. I'm inclined to keep it like this since we already have the tests.